core/src/Dscp/Crypto/MerkleTree.hs

```diff
@@ -285,15 +300,66 @@ fillEmptyMerkleTree plugs (Empty sieve) =
       where
         aux = \case
           ProofBranch sig left right ->
-              ProofBranch (coerseSig sig) <$> aux left <*> aux right
+              ProofBranch (coerceSig sig) <$> aux left <*> aux right
 
           ProofLeaf sig () ->
-              ProofLeaf (coerseSig sig) . (plugs Map.!) <$> next
+              ProofLeaf (coerceSig sig) . (plugs Map.!) <$> next
 
           ProofPruned sig ->
-              ProofPruned (coerseSig sig) <$ skip (mrSize sig)
+              ProofPruned (coerceSig sig) <$ skip (mrSize sig)
 
         next   = state $ \i -> (i,  i + 1)
         skip n = state $ \i -> ((), i + n)
 
-    coerseSig sig = error "coerseSig: 'MerkleSignature a' has 'a' inside!" <$> sig
+-- | Merkle proof with values at leaves removed and replaced via @Pruned@
+-- nodes.
+newtype EmptyMerkleProof a = EmptyProof (MerkleProof Void)
+    deriving newtype (Eq, Show, Generic, Serialise)
+
+-- | Absurd instance to derive 'Serialise' for 'EmptyMerkleProof' automatically.
+instance Serialise Void where
+    encode = absurd
+
+-- | List of indexed elements. Invariant: indices always go in ascending
+-- order.
+newtype IndexedList a = IndexedList [(LeafIndex, a)]
+    deriving newtype (Eq, Show, Generic)
+
+-- | Splits Merkle proof into signatures and data.
+separateProofAndData :: MerkleProof a -> (EmptyMerkleProof a, IndexedList a)
+separateProofAndData =
+    bimap EmptyProof (IndexedList . reverse . snd) .
+    flip runState (0, []) . go
+  where
+    go (ProofBranch s l r) = ProofBranch (coerceSig s) <$>
+        go l <*>
+        (modify' (first (+ pnSize l)) *> go r)
+    go (ProofPruned s)     = pure $ ProofPruned (coerceSig s)
+    go (ProofLeaf s a)     = pickLeaf a $> ProofPruned (coerceSig s)
```

Comment:
It looks awesome.
First, you make `EmptyMerkleTree` be equal `MerkleTree Void`, and now you automatically get only two constructors of `MerkleTree` available to be returned value here.

Reply:
@Martoon-00 actually, the whole story with merkle trees should be rewritten, because `MerkleProof`s in general should not contain inner nodes - only leaves and pruned ends; all the inner nodes should be calculated during validation. Keeping all the inner nodes is a waste of memory, since they should be recalculated during validation anyway.